assignment-1/pkg/controller/controller.go

```diff
@@ -0,0 +1,206 @@
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+type Controller struct {
+	KubernetesClientSet    *kubernetes.Clientset
+	SnapshotClientSet      snapshotClient.Interface
+	ExternalSnapshotClient externalSnapshotClient.Interface
+	CacheSynced            cache.InformerSynced
+	SnapshotLister         snapshotLister.SnapshotLister
+	SnapshotQueue          workqueue.RateLimitingInterface
+}
+
+func NewController(kubernetesClientSet *kubernetes.Clientset,
+	snapshotClientSet snapshotClient.Interface,
+	externalSnapshotClientSet externalSnapshotClient.Interface,
+	snapshotInformer snapshotInformer.SnapshotInformer) *Controller {
+	fmt.Println("starting controller")
+	c := &Controller{
+		KubernetesClientSet:    kubernetesClientSet,
+		SnapshotClientSet:      snapshotClientSet,
+		ExternalSnapshotClient: externalSnapshotClientSet,
+		CacheSynced:            snapshotInformer.Informer().HasSynced,
+		SnapshotLister:         snapshotInformer.Lister(),
+		SnapshotQueue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(), "myqueue"),
+	}
+	snapshotInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.handleAdd,
+			UpdateFunc: c.handleUpdate,
+		},
+	)
+	return c
+}
+
+func (c *Controller) handleAdd(obj interface{}) {
+	fmt.Println("handling create snapshot")
+	c.SnapshotQueue.Add(obj)
+}
+
+func (c *Controller) handleUpdate(oldObj interface{}, newObj interface{}) {
+	fmt.Println("handling update snapshot")
+	c.SnapshotQueue.Add(newObj)
+}
+
+func (c *Controller) Run(ch chan struct{}) error {
+	if ok := cache.WaitForCacheSync(ch, c.CacheSynced); !ok {
+		fmt.Println("cache was not synced")
+	}
+	go wait.Until(c.worker, time.Second, ch)
+	<-ch
+	return nil
+}
+
+func (c *Controller) worker() {
+	for c.processItems() {
+	}
+}
+
+func (c *Controller) processItems() bool {
+	item, shutdown := c.SnapshotQueue.Get()
+	if shutdown {
+		return false
+	}
+	c.SnapshotQueue.Forget(item)
+	key, err := cache.MetaNamespaceKeyFunc(item)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	snapshot, err := c.SnapshotLister.Snapshots(namespace).Get(name)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	if snapshot.Spec.Operation == "backup" {
+		err := c.createBackup(snapshot)
+		if err != nil {
+			fmt.Println(err.Error())
+			return false
+		}
+	} else if snapshot.Spec.Operation == "restore" {
+		err := c.restoreBackup(snapshot)
+		if err != nil {
+			fmt.Println(err.Error())
+			return false
+		}
+	}
+	return true
+}
+
+func (c *Controller) createBackup(snapshot *v1alpha1.Snapshot) error {
+	// PVC doesn't exists, return error
+	pvc, err := c.KubernetesClientSet.CoreV1().PersistentVolumeClaims(
+		snapshot.Namespace).Get(context.Background(), snapshot.Spec.PvcName,
```

Comment:
instead of creating context.Background(), derive it from parent as arg to the function